geth/whisper/request_historic_messages.go

```diff
@@ -70,118 +74,143 @@ func RequestHistoricMessagesHandler(nodeManager common.NodeManager) rpc.Handler
-	PoW      float64             //whisper proof of work
-}
+	// MailServer enode address.
+	Peer []byte
 
-func parseArgs(args ...interface{}) (historicMessagesRequest, error) {
-	var (
-		r = historicMessagesRequest{
-			TimeLow: uint32(time.Now().Add(-24 * time.Hour).Unix()),
-			TimeUp:  uint32(time.Now().Unix()),
-		}
-	)
+	// TimeLow is a lower bound of time range (optional).
+	// Default is 24 hours back from now.
+	TimeLow uint32
+
+	// TimeUp is a upper bound of time range (optional).
+	// Default is now.
+	TimeUp uint32
 
+	// Topic is a regular Whisper topic.
+	Topic whisperv5.TopicType
+
+	// SymKeyID is an ID of a symmetric key used to authenticate to MailServer.
+	// It's derived from MailServer password.
+	SymKeyID string
+
+	// PoW is Whisper Proof of Work value.
+	PoW float64
+}
+
+func parseArgs(args []interface{}) (r historicMessagesRequest, err error) { //nolint: gocyclo
 	if len(args) != 1 {
-		return historicMessagesRequest{}, ErrInvalidNumberOfArgs
+		return r, errInvalidNumberOfParams
 	}
 
-	historicMessagesArgs, ok := args[0].(map[string]interface{})
+	param, ok := args[0].(map[string]interface{})
 	if !ok {
-		return historicMessagesRequest{}, ErrInvalidArgs
+		return r, errInvalidParams
 	}
 
-	if t, ok := historicMessagesArgs["from"]; ok {
-		if parsed, ok := t.(uint32); ok {
-			r.TimeLow = parsed
-		}
-	}
-	if t, ok := historicMessagesArgs["to"]; ok {
-		if parsed, ok := t.(uint32); ok {
-			r.TimeUp = parsed
-		}
+	enodeInterface, ok := param["enode"]
+	if !ok {
+		return r, errMissingEnode
 	}
-	topicInterfaceValue, ok := historicMessagesArgs["topic"]
+	enode, ok := enodeInterface.(string)
 	if !ok {
-		return historicMessagesRequest{}, ErrTopicNotExist
+		return r, errInvalidEnode
 	}
+	nodeInfo, err := discover.ParseNode(enode)
+	if err != nil {
+		return r, fmt.Errorf("%v: %v", errInvalidEnode, err)
+	}
+	r.Peer = nodeInfo.ID[:]
 
-	topicStringValue, ok := topicInterfaceValue.(string)
+	topicInterface, ok := param["topic"]
 	if !ok {
-		return historicMessagesRequest{}, ErrTopicNotString
+		return r, errMissingTopic
 	}
-
-	if err := r.Topic.UnmarshalText([]byte(topicStringValue)); err != nil {
-		return historicMessagesRequest{}, nil
+	switch t := topicInterface.(type) {
+	case string:
+		r.Topic = whisperv5.BytesToTopic([]byte(t))
+	case []byte:
+		r.Topic = whisperv5.BytesToTopic(t)
+	case whisperv5.TopicType:
+		r.Topic = t
 	}
 
-	symkeyIDInterfaceValue, ok := historicMessagesArgs["symKeyID"]
+	symKeyIDInterface, ok := param["symKeyID"]
 	if !ok {
-		return historicMessagesRequest{}, ErrMailboxSymkeyIDNotExist
+		return r, errMissingSymKeyID
 	}
-	r.SymkeyID, ok = symkeyIDInterfaceValue.(string)
+	r.SymKeyID, ok = symKeyIDInterface.(string)
 	if !ok {
-		return historicMessagesRequest{}, ErrMailboxSymkeyIDNotString
+		return r, errInvalidSymKeyID
 	}
 
-	peer, err := getPeerID(historicMessagesArgs)
-	if err != nil {
-		return historicMessagesRequest{}, err
+	if t, ok := param["from"]; ok {
```

Comment:
It looks like a really complicated code (from here and up to 162L). Could we split and simplify it?

Reply:
Yeah, we could extract parsing each param to a separate function but I think we could get rid of this completely if we choose to use the node service solution.

Reply:
https://github.com/golang/go/wiki/CodeReviewComments#indent-error-flow

we could add helper:
```
func getTime(m map[string]interface{}, key string, diff time.Duration) (t uint32, err error) {
	t = uint32(time.Now().UTC().Add(diff).Unix())

	tValue, ok := m[key]
	if !ok {
		return
	}
	
	if t, ok = parseToUint32(tValue); !ok {
		return 0, fmt.Errorf("from value must be unix time in seconds, got: %T", tValue)
	}
	
	return 
}
```
and rewrite complicated if's:
```
r.TimeLow, err = getTime(param, "from", -24*time.Hour)
if err != nil {
	return r, err
}

r.TimeUp, err = getTime(param, "to", 0)
if err != nil {
	return r, err
}
```